### ISSUE TITLE
Custom iterable Niimg not handled

### ISSUE THREAD

**Issue by AlexandreAbraham** [nilearn/nilearn#349]
The documentation states that a Niimg is `any object exposing get_data() and get_affine() methods, for instance a Nifti1Image from nibabel`. This is false as some parts of the code were made with Nifti1Image in mind. For example, using an object exposing the right methods but also `__iter__()` will make `check_niimg` fail.

Example iterable object:
```python
import numpy as np
from nilearn._utils.niimg_conversions import check_niimg


class Niimg(object):

    def get_data(self):
        return np.ones((10, 10, 10, 5))

    def get_affine(self):
        return np.eye(4)

    def __iter__(self):
        for i in range(5):
            yield np.ones((10, 10, 10))

check_niimg(Niimg())
```

Commenting this code in niimg_conversions makes it work. In fact, we can't make assumptions on the data based on the presence of an `__iter__` method.
```python
    if hasattr(niimg, "__iter__"):
        pass
        #if ensure_3d:
        #    raise TypeError("A 3D image is expected, but an iterable was"
        #        " given: %s" % short_repr(niimg))
        #if hasattr(niimg, "__len__") and len(niimg) == 0:
        #    raise TypeError('An empty object - %r - was passed instead of an '
        #                    'image or a list of images' % niimg)
        #return concat_niimgs(niimg)
```
This code has to be replaced by proper checks, and we should maybe make a pass on the code to be sure that it does not happen elsewhere.

**Comment by lesteve** [comment]
Just curious, did you actually bump into this problem by reading the code and spotting this edge case or because you have your own class that has get_data, get_affine and __iter__?

Also is it just a matter of first checking for get_data and get_affine and then __iter__ rather than the other way around?

**Comment by AlexandreAbraham** [comment]
I have my custom type that I use in my scripts that supports indexing and iterating. However, I never tried to feed it to nilearn objects (as nilearn objects return most of the time Nifti1Image, I prefer to feed them filepath and then convert to my own type afterward).

**Comment by AlexandreAbraham** [comment]
Closed by #380 : we won't support custom types for the moment.
